CIP-85/README.md

```diff
@@ -0,0 +1,712 @@
+Both of these changes increase implementation complexity throughout the system: everything that processes Plutus Core terms must now handle the new cases, and handle the variable numbers of children.
+
+This change also modestly expends our novelty budget.
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
+
+This makes it different to all other Plutus Core terms, but we believe it is the expected behaviour: _no_ language has strict case destructuring.
+
+Consider e.g.
+
+```haskell
+case (Just 1) of 
+   Just x -> print "Success!"
+   Nothing -> error "Failure!"
+```
+
+Even in a strict language like Rust, nobody would expect this to evaluate the `Nothing` branch and evaluate to an error.
+
+Furthermore, this additional laziness has significant performance benefits, see Appendix 2 for more discussion.
+
+#### Is this still worth it if users aren't decoding the script context?
+
+The issues with decoding the script context only apply to users who actually try and decode it; many people have moved on to just using the underlying Data object instead.
+Arguably that saps the motivation for this proposal.
+
+There are two responses to this:
+1. This proposal seems worth it on the basis of its other benefits (benefits 2 and 3).
+2. Removing the problems with decoding the script context may allow users to return to that approach, which can be more ergonomic (and potentially cheaper, since builtin calls are expensive).
+
+#### Why not also make datums and redeemers terms instead of `Data`?
+
+It is always nice to get rid of things, and this CIP does _not_ let us get rid of `Data`, because it is still used for redeemers and datums.
+That begs the question: could we also have redeemers and datums be Plutus Core terms instead of values of type `Data`?
+
+The problem is that, unlike the encoded script context, redeemers and datums are _user facing_.
+They appear in the block format, they can be supplied by users (e.g. directly at the CLI!), and they can be inspected in block explorers.
+There is therefore a significant benefit to having them in a simpler format that is easier to work with.
+If they were Plutus Core terms in generality, we could have all kinds of strange things like people embedding functions in them.
+
+So it still seems beneficial to have redeemers and datums being `Data`.
+
+See "Alternatives: Convert Redeemers and Datums to Sums-of-Products Rather Than Builtin Data" for a variant.
+
+#### What about users who rely on the script context being `Data`?
```

Comment:
We also don't have Eq, Ord, etc. for SOPs either, since they might contain functions.
This might be a reason to maintain the status quo (even for ScriptContext), however, perhaps you could morph `Data` into something like the SOPs, giving it operations like `case` and `constr` to speed up processing.

Reply:
Been reading through this for a while. And here are some of my thoughts. Will add more later as well.

The biggest current advantage of the Script Context being Data is cheap comparisons with arbitrary data structures in the script context using equalsData. 

A good example of this is called "output tagging". Which is based on doing a txref comparison to see which outputs contain a matching datum and thus are important to the executing script. In this case equalsData allows for a non-mem intensive comparison. The hope here is that we explore all the areas of cost loss as well. 

Whether or not we convert datum/redeemer to Term as well, we still lose the cheap comparison between 2 arbitrary data objects if Script Context is changed from Data.

Reply:
You _could_ have equivalent operations on the SOPs, but they would have to fail when used on SOPs with functions in them. This isn't inherently bad, you could have it be total in PIR still in some way, but it's not nice. OTOH, doing equals etc. on SOPs should also become cheaper with this.

Reply:
What do you mean? Are you saying sum of products would have an equivalent builtin to compare 2 arbitrary constrs?

Reply:
Great, this is another example like the use of `serialiseData`. I'll add it.

I don't know if we can have a comparably fast equality on SOPs. It should be pretty fast, but probably not as fast as `equalsData` since that gets to shell out to Haskell. I'll construct another benchmark to test this.

Reply:
@michaelpj I think that every single "alternative solution" outside of PlutusTx today (i.e, Aiken, Helios, Plu-ts, Hebi (ex Eopsin)) is relying on that "accidental feature". 

To be honest, the situation isn't to bad when relying on that. I think most projects went a bit with the same approach and looked at what was available on-chain, and walk from there from the bottom up. Whereas PlutusTx has had a more top-down approach going from Haskell and trying to fit that in the Plutus VM.

Reply:
Well, the tradeoff would be that you get faster "normal" processing of script context objects (i.e. any user-defined function that's pattern matching or whatever will be faster), but slower processing for the two "bulk" operations that can currently be done with builtins.

Reply:
Haxe a subtype of SOPs without functions in TPLC where an equality builtin could work? Could otherwise be the same in UPLC.

Reply:
You can call this other form of SOPs without functions "Data".